packages/react-reconciler/src/ReactFiberWorkLoop.new.js

```diff
@@ -3184,9 +3185,12 @@ function doubleInvokeEffectsInDEV(
 ) {
   const isStrictModeFiber = fiber.type === REACT_STRICT_MODE_TYPE;
   const isInStrictMode = parentIsInStrictMode || isStrictModeFiber;
+
   if (fiber.flags & PlacementDEV || fiber.tag === OffscreenComponent) {
     setCurrentDebugFiberInDEV(fiber);
-    if (isInStrictMode) {
+    const hasOffscreenVisibilityFlag =
+      fiber.tag !== OffscreenComponent || fiber.flags & Visibility;
+    if (isInStrictMode && hasOffscreenVisibilityFlag) {
```

Comment:
Can you explain why this fixes it? Is it because we shouldn't fire effects in a hidden offscreen tree?

Reply:
As far as I understand, Visibility flag is set on Offscreen when its visibility changes.

We only want to fire effects on Offscreen when it is visible and only when its visibility changes. This prevents the infinite loop. Without it, it kept firing effects unconditionally if any effect causes synchronous re-render.

Reply:
I tried running the regression test on main and it passes even without any changes, so we need to fix that

Reply:
@acdlite if you think we don't need special logic for Offscreen, I'm happy to change the tests. But I think it will be rather confusing for our users.

Reply:
> I think it would be unexpected for our users to see mount/unmount on a component that is hidden by Offscreen.

Hm maybe I misunderstand, but if we don't go through the mount/unmount in hidden trees, how will users know if there are issues in those trees exposed by the strict effects behavior?

Reply:
They will need to trigger logic in their code that will make Offscreen visible. 
I think you can make the argument one way or the other. When I was thinking about it, to me it made more sense to only fire effects on visible components. I can see the value of firing them always, even on hidden subtrees. I'm happy to change it however we see fit.

Reply:
I see. I think the behavior that makes sense to me. It would double a tree whenever it's revealed, whether that's via a regular mount or an Offscreen becoming visible.

Aside from checking just the Visibility flag, though, you should also check if it's being revealed or hidden, since it gets set in both cases. `offscreenFiber.memoizedState === null` means it's being revealed.

Reply:
This is now changed with tests included.

Reply:
What if you change code in a parent that would break an offscreen subtree, but you don't know because you don't navigate to that offscreen component in development (because you're not working on that tab/modal, for example). It seems like these type of bugs are something you eagerly want to know about with StrictMode.

Curious what @gaearon thinks as well.

Reply:
> What if you change code in a parent that would break an offscreen subtree, but you don't know because you don't navigate to that offscreen component in development (because you're not working on that tab/modal, for example).

I agree, there is definitely a value in double invoking effects even if Offscreen is hidden. On the other hand, it will cause confusion. Suddenly you would see effects triggering for something you wouldn't expect. This will be even more confusing with pre rendering where a screen that engineer might not even know about has its effects executed. I think there is already expectation with StrictMode that component needs to be used to trigger double invoking. It aligns well with Offscreen suppressing when hidden.

Reply:
> Suddenly you would see effects triggering for something you wouldn't expect. 

Isn't this also the case for double rendering? Is it confusing that we double render those hidden trees but don't double fire the effects?

Reply:
I don't think so. Take pre-rendering as an example. User expects render functions to be executed even if the subtree is hidden. This is true for pre-rendering example or for possible virtualised list implementations.
I do see value in double invoking effects in hidden subtree as well. I don't hold strong opinions on this. I'm just explaining how I was thinking about this when I implemented it.

Reply:
Discussed offline, the other concern here is that to do this right, you also need to hide the currently visible content, as though you fully switch to the hidden tab. And when you switch to the hidden tab, you may not have the props you need to do a proper render. For example, if you're getting the `userId` from the route, then you don't have an ID to test the offscreen tree with, so you end up testing behavior that doesn't occur in the wild. 

So it's easier to test making visible content hidden than it is testing hidden/pre-rendered content visible without more information. The best practice is to test navigating to the offscreen trees.